Add style guide and basic contribution section (#3651)
open-telemetry/opentelemetry.io: content/en/docs/Contribute/_index.md

Author: TIL I needed two git mv commands to do that :) https://stackoverflow.com/a/11183844

---
title: Contribute
description: Learn how to contribute to the OpenTelemetry documentation.
weight: 200
cSpell:ignore: linktitle open-telemetry prepopulated repos spacewhite
---

You can open an issue about OpenTelemetry documentation, or contribute a change
with a pull request (PR) to the
[`opentelemetry.io` GitHub repository](https://github.com/open-telemetry/opentelemetry.io).

OpenTelemetry documentation contributors:

- Improve existing content.
- Create new content.
- Update the OpenTelemetry Registry.
- Improve the code that builds the site.

See also the general
[OpenTelemetry Contributor Guide](https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md)
, which provides details on the Contributor License Agreement and the Code of
Conduct.

## Requirements

To contribute, you need to be familiar with the following techs and tools:

> - git
> - GitHub
> - Markdown (CommonMark)
> - YAML

Reviewer: ```suggestion
- [git](https://git-scm.com/)
- [GitHub](https://github.com/)
- Markdown ([CommonMark](https://commonmark.org/))
- YAML
```

For technical details concerning how the documentation is built and tested
locally, see the
[CONTRIBUTING.md](https://github.com/open-telemetry/opentelemetry.io/blob/main/CONTRIBUTING.md)
file.

### Sign the CNCF CLA {#sign-the-cla}

All OpenTelemetry contributors must read the
[Contributor guide](https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md)
and
[sign the Contributor License Agreement (CLA)](https://docs.linuxfoundation.org/lfx/easycla/contributors)
.

Pull requests from contributors who haven't signed the CLA fail the automated
tests. The name and email you provide must match those found in your
`git config`, and your git name and email must match those used for the CNCF
CLA.

## Contribute new content

```mermaid
flowchart LR
    subgraph first[How to contribute]
    direction TB
       T[ ] -.-
       A[Sign the CNCF CLA] --> D[Write docs in markdown<br>and build site with Hugo]
       D[Write docs in markdown<br>and build site with Hugo] --- E[Push source to GitHub]

>        E --- G[Open a pull request]

Reviewer: Signing the CLA is 95% of the time something that happens AFTER the pull request has been opened. Also, I miss the piece around forking the repo and pushing to the fork first and then opening the PR

    2 --> 3[3. fill in Propose file change]

end
subgraph tasks2[ ]
direction TB
4[4. select Propose file change] --> 5[5. select Create pull request] --> 6[6. fill in Open a pull request]
6 --> 7[7. select Create pull request]
end

id1 --> tasks --> tasks2

classDef grey fill:#dddddd,stroke:#ffffff,stroke-width:px,color:#000000, font-size:15px;
classDef white fill:#ffffff,stroke:#000,stroke-width:px,color:#000,font-weight:bold
classDef k8s fill:#326ce5,stroke:#fff,stroke-width:1px,color:#fff;
classDef spacewhite fill:#ffffff,stroke:#fff,stroke-width:0px,color:#000
class A,1,2,3,4,5,6,7 grey
class 0 spacewhite
class tasks,tasks2 white
class id1 k8s
```

Figure 1. Steps for opening a PR using GitHub.

1. On the page where you see the issue, select the **Edit this page** option in
   the right-hand side navigation panel.

1. Make your changes in the GitHub editor.

1. Below the editor, fill in the **Propose file change** form.

1. Select **Propose file change**.

1. Select **Create pull request**.

1. The **Open a pull request** screen appears. Your description helps reviewers
   understand your change.

> 1. Select **Create pull request**.

Reviewer: I am very certain that this flow also does not work if someone is not member of the otel gh org, there is a step where github will ask you about forking.

Author: GH implicitly creates a fork when doing a small edit from the UI. I'll add that.

> Before merging a pull request, OpenTelemetry community members review and
> approve it. If you have someone in mind, leave a comment with their GitHub
> username in it.

Reviewer: ```suggestion
Before merging a pull request, OpenTelemetry community members review and
approve it.
```

I don't think this is a good idea to call this out specifically.

If a reviewer asks you to make changes:

1. Go to the **Files changed** tab.
1. Select the pencil (edit) icon on any files changed by the pull request.
1. Make the changes requested.

> 1. Commit the changes.

Reviewer: Should we have a few words around suggestions and that you can use them to "edit" a file as well?

Author: Added!

> When your review is complete, a reviewer merges your PR and your changes go liv

Reviewer: ```suggestion
When your review is complete, a reviewer merges your PR and your changes goes live
```

   repository.
1. Select **Fork**.

#### Create a local clone and set the upstream

1. In a terminal window, clone your fork and install the requirements:

   ```shell
   git clone git@github.com:<your_github_username>/opentelemetry.io.git
   cd opentelemetry.io
   npm install
   ```

1. Set the `open-telemetry/opentelemetry.io` repository as the `upstream`
   remote:

   ```shell
   git remote add upstream https://github.com/open-telemetry/opentelemetry.io.git
   ```

1. Confirm your `origin` and `upstream` repositories:

   ```shell
   git remote -v
   ```

   Output is similar to:

   ```none
   origin	git@github.com:<your_github_username>/opentelemetry.io.git (fetch)
   origin	git@github.com:<your_github_username>/opentelemetry.io.git (push)
   upstream	https://github.com/open-telemetry/opentelemetry.io.git (fetch)
   upstream	https://github.com/open-telemetry/opentelemetry.io.git (push)
   ```

1. Fetch commits from your fork's `origin/main` and
   `open-telemetry/opentelemetry.io`'s `upstream/main`:

   ```shell
   git fetch origin
   git fetch upstream
   ```

   This makes sure your local repository is up to date before you start making

>    changes.

Reviewer: You also need to push changes to upstream to origin. 

Personally, I stopped working that way, I _rarely_ add an upstream, I keep my fork up-to-date (either manually or with an github app, etc.) and only pull from the fork. PRs are still send to upstream.

Author: Hmm, how do you keep the fork in sync with upstream if git doesn't know where to pull?

   ```

1. Make your changes using a code or text editor.

At any time, use the `git status` command to see what files you've changed.

#### Commit your changes

When you are ready to submit a pull request, commit your changes.

1. In your local repository, check which files you need to commit:

   ```shell
   git status
   ```

   Output is similar to:

   ```none
   On branch <my_new_branch>
   Your branch is up to date with 'origin/<my_new_branch>'.

   Changes not staged for commit:
   (use "git add <file>..." to update what will be committed)
   (use "git checkout -- <file>..." to discard changes in working directory)

   modified:   content/en/docs/contribute/new-content/contributing-content.md

   no changes added to commit (use "git add" and/or "git commit -a")
   ```

1. Add the files listed under **Changes not staged for commit** to the commit:

   ```shell
   git add <your_file_name>
   ```

   Repeat this for each file.

1. After adding all the files, create a commit:

   ```shell
   git commit -m "Your commit message"
   ```

1. Push your local branch and its new commit to your remote fork:

   ```shell
   git push origin <my_new_branch>

>    ```

Reviewer: GitHub these days sends back a message that you can use to create a PR from here, do we want to call this out?

1. From the **compare** drop-down menu, select your branch.
1. Select **Create Pull Request**.
1. Add a description for your pull request:

   - **Title** (50 characters or less): Summarize the intent of the change.
   - **Description**: Describe the change in more detail.

     - If there is a related GitHub issue, include `Fixes #12345` or
       `Closes #12345` in the description. GitHub's automation closes the
       mentioned issue after merging the PR if used. If there are other related
       PRs, link those as well.
     - If you want advice on something specific, include any questions you'd
       like reviewers to think about in your description.

1. Select the **Create pull request** button.

Your pull request is available in
[Pull requests](https://github.com/open-telemetry/opentelemetry.io/pulls).

After opening a PR, GitHub runs automated tests and tries to deploy a preview
using [Netlify](https://www.netlify.com/).

- If the Netlify build fails, select **Details** for more information.
- If the Netlify build succeeds, select **Details** opens a staged version of
  the OpenTelemetry website with your changes applied. This is how reviewers

>   check your changes.

Reviewer: this needs also to speak about the other PR actions that may fail

GitHub also automatically assigns labels to a PR to help reviewers.

#### Changes from reviewers

Sometimes reviewers commit to your pull request. Before making any other
changes, fetch those commits.

1. Fetch commits from your remote fork and rebase your working branch:

   ```shell
   git fetch origin
   git rebase origin/<your-branch-name>
   ```

1. After rebasing, force-push new changes to your fork:

   ```shell
   git push --force-with-lease origin <your-branch-name>

>    ```

Reviewer: I don't know, for me a simple `git pull` while being on the branch for the PR does the job, that sounds overly complicated (I never heard about `--force-with-lease`)

Author: What's your pull conflict strategy? Rebase or merge? Perhaps it's needed when merging?

#### Merge conflicts and rebasing

If another contributor commits changes to the same file in another PR, it can

> create a merge conflict. You must resolve all merge conflicts in your PR.

Reviewer: There is also an option to fix merge conflicts via the GitHub UI, this might be worth calling out.

1. Add the files to the changeset:

   ```shell
   git add <filename>
   ```

1. Continue the rebase:

   ```shell
   git rebase --continue
   ```

1. Repeat steps 2 to 5 as needed.

   After applying all commits, the `git status` command shows that the rebase is
   complete.

1. Force-push the branch to your fork:

   ```shell
   git push --force-with-lease origin <your-branch-name>
   ```

   The pull request no longer shows any conflicts.

## Contribute to other repositories

The OpenTelemetry project contains many repositories. Many of these repositories
contain documentation: user-facing help text, error messages, API references or
code comments.

If you see text you'd like to improve, use GitHub to search all repositories in
the OpenTelemetry organization. This can help you figure out where to submit
your issue or PR.

Each repository has its own processes and procedures. Before you file an issue
or submit a PR, read that repository's `README.md`, `CONTRIBUTING.md`, and
`code-of-conduct.md`, if they exist.

Most repositories use issue and PR templates. Have a look through some open
issues and PRs to get a feel for that team's processes. Make sure to fill out

> the templates with as much detail as possible when you file issues or PRs.

Reviewer: this should be way down after opening an issue

Author: Moved!

## Open an issue

If you want to suggest improvements to existing content or notice an error, open
an issue.

> 1. Click the **Create documentation issue** link on the right sidebar. This

Reviewer: Where on which right sidebar?

Author: Changed to "on any document" — it's the third link in the floating upper-right panel, together with "View page source" and "Edit this page".

   redirects you to a GitHub issue page prepopulated with some headers.

> 2. Describe the issue or suggestion for improvement. Provide as many details as

Reviewer: we have templates, so ask people to pick the right one.

Author: Ah, I see what you mean! Edited both flows

   you can.
3. Click **Submit new issue**.

After submitting, check in on your issue occasionally or turn on GitHub
notifications. Reviewers and other community members might ask questions before

> they can take action on your issue.

Reviewer: Add a few words about patience, that it can take a few days until maintainers respond. If they don't get an answer within 7 days they can bump the issue up with a message or reach out via slack

### Suggesting new content

If you have an idea for new content, but you aren't sure where it should go, you
can still file an issue. Either:

- Choose an existing page in the section you think the content belongs in and
  click **Create documentation issue**.
- Go to [GitHub](https://github.com/open-telemetry/opentelemetry.io/issues/new/)

>   and file the issue directly.

Reviewer: Same here, the flow is different.

Author: Is it? I think we've that link in Docsy.
